beacon_node/http_api/src/publish_blocks.rs

```diff
@@ -119,130 +144,252 @@ pub async fn publish_block<T: BeaconChainTypes, B: IntoGossipVerifiedBlockConten
-                    "slot" => slot,
-                    "error" => %e
-                );
-                return Err(warp_utils::reject::custom_bad_request(e.to_string()));
-            }
-        };
 
-    // Clone here, so we can take advantage of the `Arc`. The block in `BlockContents` is not,
-    // `Arc`'d but blobs are.
-    let block = gossip_verified_block.block.block_cloned();
-    let blobs_opt = gossip_verified_blobs.as_ref().map(|gossip_verified_blobs| {
-        let blobs = gossip_verified_blobs
+    // Check blob inclusion proofs and convert to blob sidecars ready for publication.
+    let mut blob_sidecars = if let Some((kzg_proofs, blobs)) = unverified_blobs {
+        kzg_proofs
             .into_iter()
-            .map(|b| b.clone_blob())
-            .collect::<Vec<_>>();
-        VariableList::from(blobs)
-    });
+            .zip(blobs)
+            .enumerate()
+            .map(|(i, (proof, unverified_blob))| {
+                let _timer = metrics::start_timer(
+                    &beacon_chain::metrics::BLOB_SIDECAR_INCLUSION_PROOF_COMPUTATION,
+                );
+                let blob_sidecar =
+                    BlobSidecar::new(i, unverified_blob, &block, proof).map(Arc::new);
+                blob_sidecar.map_err(|e| {
+                    error!(
+                        log,
+                        "Invalid blob - not publishing block";
+                        "error" => ?e,
+                        "blob_index" => i,
+                        "slot" => slot,
+                    );
+                    warp_utils::reject::custom_bad_request(format!("{e:?}"))
+                })
+            })
+            .collect::<Result<Vec<_>, Rejection>>()?
+    } else {
+        vec![]
+    };
 
-    let block_root = block_root.unwrap_or(gossip_verified_block.block_root);
+    // Gossip verify the block and blobs separately.
+    let gossip_verified_block_result = unverified_block.into_gossip_verified_block(&chain);
+    let gossip_verified_blobs = blob_sidecars
+        .iter_mut()
+        .map(|blob_sidecar| {
+            let gossip_verified_blob =
+                GossipVerifiedBlob::new(blob_sidecar.clone(), blob_sidecar.index, &chain);
+
+            match gossip_verified_blob {
+                Ok(blob) => Ok(Some(blob)),
+                Err(GossipBlobError::RepeatBlob { proposer, .. }) => {
+                    // Log the error but do not abort publication, we may need to publish the block
+                    // or some of the other blobs if the block & blobs are only partially published
+                    // by the other publisher.
+                    debug!(
+                        log,
+                        "Blob for publication already known";
+                        "blob_index" => blob_sidecar.index,
+                        "slot" => slot,
+                        "proposer" => proposer,
+                    );
+                    Ok(None)
+                }
```

Comment:
@jimmygchen I think this might handle the case where a block is being imported from gossip and HTTP, and both imports trip the `RepeatBlob` condition.

With this change, the HTTP API will continue with trying to import the block.

If the block is also seen on gossip before HTTP we will get a `DuplicateImportStatusUnknown` and still call `process_block`

Reply:
Yep that looks right!

Reply:
So i think our issue is we're re-processing the block here:
https://github.com/sigp/lighthouse/blob/4f3cf2c7292a127c76cc7c81063fcb408fc7a1d7/beacon_node/http_api/src/publish_blocks.rs#L336-L343

Which is already imported earlier via `process_gossip_blob`
https://github.com/sigp/lighthouse/blob/4f3cf2c7292a127c76cc7c81063fcb408fc7a1d7/beacon_node/http_api/src/publish_blocks.rs#L270-L273

And because of the `BlockError`, the head is not recomputed in `post_block_import_logging_and_response`. We should probably just use the `AvailabilityProcessingStatus` from last `process_gossip_blob` call? and if the block is  imported it will recompute the head.

Reply:
Logs and some notes here:

```
# [gossip] block received
2024-07-29T05:22:23.1173355Z Jul 29 05:22:23.117 INFO New block received                      root: 0xa37555b6829209e977de1a8ed8a40651e4536e92d0c7c39bdeb81e162749dea8, slot: 43, service: node_1

# [gossip] blob 0 received
2024-07-29T05:22:23.1201914Z Jul 29 05:22:23.120 DEBG Successfully verified gossip blob       commitment: 0xa41e…f05f, index: 0, root: 0xa375…dea8, slot: 43, service: node_1
2024-07-29T05:22:23.1378924Z Jul 29 05:22:23.137 DEBG Beacon block awaiting blobs             block_slot: 43, block_root: 0xa37555b6829209e977de1a8ed8a40651e4536e92d0c7c39bdeb81e162749dea8, service: beacon, service: node_1

# [gossip] blob 1 received
2024-07-29T05:22:23.1499531Z Jul 29 05:22:23.149 DEBG Successfully verified gossip blob       commitment: 0xa41e…f05f, index: 1, root: 0xa375…dea8, slot: 43, service: node_1

# [http] block and blobs (0,1,2,3) received
2024-07-29T05:22:23.1590330Z Jul 29 05:22:23.158 DEBG Signed block received in HTTP API       slot: 43, service: node_1

# [http] blob 0 already known from gossip
2024-07-29T05:22:23.1621330Z Jul 29 05:22:23.161 DEBG Blob for publication already known      proposer: 9, slot: 43, blob_index: 0, service: node_1

# [http] blob 1 already known from gossip
2024-07-29T05:22:23.1623295Z Jul 29 05:22:23.162 DEBG Blob for publication already known      proposer: 9, slot: 43, blob_index: 1, service: node_1

# [gossip] blob 2 already known from http
2024-07-29T05:22:23.1742409Z Jul 29 05:22:23.171 DEBG Could not verify blob sidecar for gossip. Ignoring the blob sidecar, commitment: 0xa41e…f05f, index: 2, root: 0xa375…dea8, slot: 43, error: RepeatBlob { proposer: 9, slot: Slot(43), index: 2 }, service: node_1

# [http] nothing is actually published here it seems, because of the import below, and gossip block received beforehand, so `should_import` = false. We might want to remove this log?
2024-07-29T05:22:23.1783377Z Jul 29 05:22:23.178 INFO Signed block published to network via HTTP API, publish_delay_ms: 19, slot: 43, service: node_1

# [http] we're in `chain.import_block`! (possibly via `chain.process_gossip_blob` in `publish_block`)
2024-07-29T05:22:23.1816196Z Jul 29 05:22:23.181 DEBG Writing blobs to store                  count: 4, block_root: 0xa375…dea8, service: beacon, service: node_1

# [gossip] blob 3 already known from http
2024-07-29T05:22:23.1838572Z Jul 29 05:22:23.181 DEBG Could not verify blob sidecar for gossip. Ignoring the blob sidecar, commitment: 0xa41e…f05f, index: 3, root: 0xa375…dea8, slot: 43, error: RepeatBlob { proposer: 9, slot: Slot(43), index: 3 }, service: node_1

# [http] api partially publishing blobs in `chain.process_gossip_blob`
2024-07-29T05:22:23.1865775Z Jul 29 05:22:23.185 DEBG Sending pubsub messages                 topics: [BlobSidecar(2), BlobSidecar(3)], count: 2, service: network, service: node_1

# [http] block seen from gossip, import status unknown, try processing the block. 
2024-07-29T05:22:23.1902460Z Jul 29 05:22:23.186 DEBG Block previously seen                   slot: 43, block_root: 0xa37555b6829209e977de1a8ed8a40651e4536e92d0c7c39bdeb81e162749dea8, service: node_1

# [http] rejected. We've already imported the block in `chain.process_gossip_blob` and we should probably look at the last `AvailabilityProcessingStatus` from that call instead of looking at an outdated process_block result?
2024-07-29T05:22:23.1905042Z Jul 29 05:22:23.187 DEBG Beacon block rejected                   reason: DuplicateFullyImported(0xa37555b6829209e977de1a8ed8a40651e4536e92d0c7c39bdeb81e162749dea8), service: beacon, service: node_1
2024-07-29T05:22:23.1907689Z Jul 29 05:22:23.187 WARN Error processing HTTP API request       method: POST, path: /eth/v1/beacon/blocks, status: 202 Accepted, elapsed: 54.184048ms, service: node_1
2024-07-29T05:22:24.0035956Z Jul 29 05:22:24.001 DEBG Slot timer                              sync_state: Synced, current_slot: 43, head_slot: 42, head_block: 0x2c26…f8b3, finalized_epoch: 3, finalized_root: 0x83ad…d15a, peers: 5, service: slot_notifier, service: node_1

# block is not yet head. That's expected because we've imported block but haven't recomputed head.
2024-07-29T05:22:24.0037668Z Jul 29 05:22:24.001 INFO Synced                                  slot: 43, block:    …  empty, epoch: 5, finalized_epoch: 3, finalized_root: 0x83ad…d15a, exec_hash: 0x694a…b4c0 (verified), peers: 5, service: slot_notifier, service: node_1
1918, slot: 43, proposer_index: 9, block_root: 0xa37555b6829209e977de1a8ed8a40651e4536e92d0c7c39bdeb81e162749dea8, service: beacon, service: node_1
```

Reply:
Resolved this case in d37b0806d948cb15f4317ad923a0ba70cab28ed1.

I am going to try soak testing the simulator with this branch.